src/wms/src/main/java/org/geoserver/wms/legendgraphic/LegendMerger.java

```diff
@@ -418,13 +419,111 @@ public int getHeight() {
     }
 
     /**
-     * Creates legends columns for vertical layout according to max height and max columns limits
+     * Creates non-raster legend columns for vertical layout according to max height and max columns
+     * limits
      * 
      * @param nodes legend images
      * @param maxHeight maximum height of legend
      * @param maxColumns maximum number of columns
+     * @param req general mechanism for acquiring legend symbols
      * 
      */
+    private static Column[] createNRColumns(List<BufferedImage> nodes, int maxHeight, int maxColumns,
+            GetLegendGraphicRequest req) {
+        Column[] legendMatrix = new Column[0];
+        /*
+         * Limit max height
+         */
+        if (maxHeight > 0) {
+            /*
+             * Limit max column
+             */
+            int cnLimit = maxColumns > 0 ? maxColumns : nodes.size();
+            legendMatrix = new Column[cnLimit];
+            legendMatrix[0] = new Column();
+            int cn = 0;
+            int columnHeight = 0;
+            for (int i = 0; i < nodes.size(); i++) {
+                BufferedImage node = nodes.get(i);
+                if (columnHeight <= maxHeight) {
+                    // Fill current column
+                    legendMatrix[cn].addNode(node);
+                    columnHeight = columnHeight + node.getHeight();
+                } else {
+                    // Add current node to next column
+                    i--;
+                    cn++;
+                    // Stop if column limits is reached
+                    if (cn == cnLimit) {
+                        break;
+                    }
+                    // Reset column counter
+                    columnHeight = 0;
+                    // Create new column
+                    legendMatrix[cn] = new Column();
+                }
+            }
+        } else {
+            /*
+             * Limit max column, if no limit set it to 1
+             */
+            int colNumber = maxColumns > 0 ? maxColumns : 1;
+            legendMatrix = new Column[colNumber];
+            legendMatrix[0] = new Column();
+            int rowNumber = (int) Math.ceil((float) nodes.size() / colNumber);
+            int cn = 0;
+            int rc = 0;
+            boolean colourPresent = false;
+            for (int i = 0; i < nodes.size(); i++) {
+                if (rc < rowNumber) {
+                    // check for presence of colour (ie. non-empty legend row)
+                    colourPresent = checkColour(nodes.get(i), req);
+                    if(colourPresent){
+                        legendMatrix[cn].addNode(nodes.get(i));
+                        rc++;
+                    }
+                } else {
+                    i--;
+                    cn++;
+                    rc = 0;
+                    legendMatrix[cn] = new Column();
+                }
+            }
+        }
+
+        return legendMatrix;
+    }

```

Comment:
The vast majority of this method is identical to createColumns. Please refactor to avoid duplicated code.